File: problems/0355-design-twitter/go/solution.go
package solution

import "sort"

// Design Twitter — https://leetcode.com/problems/design-twitter/
type tweet struct {
	time int
	id   int
}

type Twitter struct {
	time      int
	tweets    map[int][]tweet
	following map[int]map[int]bool
}

func NewTwitter() *Twitter {
	return &Twitter{tweets: map[int][]tweet{}, following: map[int]map[int]bool{}}
}

func (t *Twitter) PostTweet(userId, tweetId int) {
	t.tweets[userId] = append(t.tweets[userId], tweet{t.time, tweetId})
	t.time++
}

func (t *Twitter) GetNewsFeed(userId int) []int {
	all := append([]tweet(nil), t.tweets[userId]...)
	for f := range t.following[userId] {
		if f != userId {
			all = append(all, t.tweets[f]...)
		}
	}
	sort.Slice(all, func(i, j int) bool { return all[i].time > all[j].time })
	res := []int{}
	for i := 0; i < len(all) && i < 10; i++ {
		res = append(res, all[i].id)
	}
	return res
}

func (t *Twitter) Follow(followerId, followeeId int) {
	if t.following[followerId] == nil {
		t.following[followerId] = map[int]bool{}
	}
	t.following[followerId][followeeId] = true
}

func (t *Twitter) Unfollow(followerId, followeeId int) {
	if t.following[followerId] != nil {
		delete(t.following[followerId], followeeId)
	}
}

func twitterOps(ops []string, args [][]int) [][]int {
	tw := NewTwitter()
	out := [][]int{}
	for i, op := range ops {
		switch op {
		case "Twitter":
			tw = NewTwitter()
		case "postTweet":
			tw.PostTweet(args[i][0], args[i][1])
		case "getNewsFeed":
			out = append(out, tw.GetNewsFeed(args[i][0]))
		case "follow":
			tw.Follow(args[i][0], args[i][1])
		case "unfollow":
			tw.Unfollow(args[i][0], args[i][1])
		}
	}
	return out
}

File: problems/0355-design-twitter/typescript/solution.ts
// Design Twitter — https://leetcode.com/problems/design-twitter/
class Twitter {
  private time = 0;
  private tweets = new Map<number, { time: number; id: number }[]>();
  private following = new Map<number, Set<number>>();

  postTweet(userId: number, tweetId: number): void {
    const list = this.tweets.get(userId) ?? [];
    list.push({ time: this.time++, id: tweetId });
    this.tweets.set(userId, list);
  }

  getNewsFeed(userId: number): number[] {
    const all = [...(this.tweets.get(userId) ?? [])];
    const followees = this.following.get(userId);
    if (followees) {
      for (const f of followees) {
        if (f !== userId) all.push(...(this.tweets.get(f) ?? []));
      }
    }
    all.sort((a, b) => b.time - a.time);
    return all.slice(0, 10).map((t) => t.id);
  }

  follow(followerId: number, followeeId: number): void {
    const set = this.following.get(followerId) ?? new Set<number>();
    set.add(followeeId);
    this.following.set(followerId, set);
  }

  unfollow(followerId: number, followeeId: number): void {
    this.following.get(followerId)?.delete(followeeId);
  }
}

export function twitterOps(ops: string[], args: number[][]): number[][] {
  let tw = new Twitter();
  const out: number[][] = [];
  for (let i = 0; i < ops.length; i++) {
    const a = args[i];
    switch (ops[i]) {
      case "Twitter":
        tw = new Twitter();
        break;
      case "postTweet":
        tw.postTweet(a[0], a[1]);
        break;
      case "getNewsFeed":
        out.push(tw.getNewsFeed(a[0]));
        break;
      case "follow":
        tw.follow(a[0], a[1]);
        break;
      case "unfollow":
        tw.unfollow(a[0], a[1]);
        break;
    }
  }
  return out;
}
